### PR TITLE
Fix attribute key encoding

### PR DESCRIPTION
**Pull request by es-sayers**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  |

Attribute keys are currently not encoded. Since `Tag` implements `NoEncodeStringableInterface`, HTML characters may be output unintentionally.

### CODE REVIEW

**Review by vjik — COMMENTED** [review]
I think, encode attribute name is bad idea, because as result will be another name. 

[HTML standard](https://html.spec.whatwg.org/multipage/syntax.html#attributes-2):

> Attributes have a name and a value. Attribute names must consist of one or more characters other than [controls](https://infra.spec.whatwg.org/#control), U+0020 SPACE, U+0022 ("), U+0027 ('), U+003E (>), U+002F (/), U+003D (=), and [noncharacters](https://infra.spec.whatwg.org/#noncharacter). In the HTML syntax, attribute names, even those for [foreign elements](https://html.spec.whatwg.org/multipage/syntax.html#foreign-elements), may be written with any mix of [ASCII lower](https://infra.spec.whatwg.org/#ascii-lower-alpha) and [ASCII upper alphas](https://infra.spec.whatwg.org/#ascii-upper-alpha).

Removing denied symbols also looks bad.

I suggest don't modify attribute names and let the developer solve this problem himself.

@samdark What do you think?

**Comment by samdark** [comment]
I think that these should be escaped/encoded for security reasons. It doesn't make sense if there are forbidden characters anyway.

**Comment by samdark** [comment]
@es-sayers would you please add a line to CHANGELOG?

**Comment by vjik** [comment]
> I think that these should be escaped/encoded for security reasons. It doesn't make sense if there are forbidden characters anyway.

This is not `escaped/encoded`. This code change name to another (`>hello` → `&gt;hello`). It's incorrect, browser don't unescape attribute names.

Sanitizing (remove denied symbols) is better for security. We can make new method `Html::sanitizeAttributeName()` for it.

**Comment by samdark** [comment]
> browser don't unescape attribute names

That's false.

```html
<form method="GET">
  <label id="first">
    Field
    <input type="text" name="&gt;hello" required />
  </label>
  <input type="submit" name="save" value="Save" />
</form>
```

Try yourself. Backend will get `>hello` as GET name.

**Comment by vjik** [comment]
>     <input type="text" name="&gt;hello" required />

This PR about attribute key, not attribute value.

**Comment by es-sayers** [comment]
>Attribute names must consist of one or more characters other than [controls](https://infra.spec.whatwg.org/#control), U+0020 SPACE, U+0022 ("), U+0027 ('), U+003E (>), U+002F (/), U+003D (=), and [noncharacters](https://infra.spec.whatwg.org/#noncharacter).

Since attribute names can't have control characters, I think it's best to remove the vulnerability and have the developer work around it.

**Comment by es-sayers** [comment]
Is there any alternative behavior that would be better?

With encoding, a dev might try to have `>custom="value"` and receive `&gt;custom="value"`. Which is a valid attribute but not what they wanted.

With removing control characters, a dev might have `attribute>1="value"` and `attribute<1="value2"` which would both return as `attribute1="..."`

I think it might be better to just fully remove the invalid attribute while still rendering the rest of the attribute string. I think this is the only way to limit unexpected behavior while also removing the security risk.

Other than that, throwing `InvalidArgumentException` might apply here, similarly to how it is applied in `Range::outputTag()`.

```php
    public function outputTag(string $tagName): self
    {
        if ($tagName === '') {
            throw new InvalidArgumentException('The output tag name it cannot be empty value.');
        }

        $new = clone $this;
        $new->outputTag = $tagName;
        return $new;
    }
```

**Comment by samdark** [comment]
@vjik thanks for pointing that out. @es-sayers throwing an exception might be a better option.

File: src/Html.php [hunk 1]
```diff
@@ -1956,9 +1957,16 @@ public static function getNonArrayableName(string $name): string
      * Render attribute in HTML tag.
      *
      * @link https://html.spec.whatwg.org/#a-quick-introduction-to-html
+     * @throws ValueError When name contains any of the following: SPACE " ' > / =
      */
     private static function renderAttribute(string $name, string $encodedValue = '', string $quote = '"'): string
     {
+        if (preg_match('/[\s"\'>\/\=]/', $name) > 0) {
```
Comment (samdark) [inline src/Html.php:1964]: Better to use `str_contains` instead of regular expression for this case.

File: src/Html.php [hunk 2]
```diff
@@ -1956,9 +1957,16 @@ public static function getNonArrayableName(string $name): string
...
+            throw new ValueError(
```
Comment (samdark) [inline src/Html.php:1965]: We were using `InvalidArgumentException` for such cases previously. I think it's better to be consistent.

**Review by samdark — CHANGES_REQUESTED** [review]

File: src/Html.php (hunk 2 above)
Comment (vjik) [reply src/Html.php:1965]: Does this regexp cover all deny characters from docs: https://html.spec.whatwg.org/multipage/syntax.html#attributes-2 ?

File: src/Html.php (hunk 1 above)
Comment (samdark) [reply src/Html.php:1964]: @vjik nope.

Comment (es-sayers) [reply src/Html.php:1963]: Do you also want checks for [https://infra.spec.whatwg.org/#noncharacter](url) and [https://infra.spec.whatwg.org/#control](url) ?

Comment (vjik) [reply src/Html.php:1963]: That would be good.

**Review by vjik — APPROVED** [review]

File: src/Html.php [hunk 3]
```diff
@@ -1956,9 +1956,37 @@ public static function getNonArrayableName(string $name): string
...
+     *     U+6FFFE, U+6FFFF, U+7FFFE, U+7FFFF, U+8FFFE, U+8FFFF, U+9FFFE, U+9FFFF, U+AFFFE, U+AFFFF, U+BFFFE,
+     *     U+BFFFF, U+CFFFE, U+CFFFF, U+DFFFE, U+DFFFF, U+EFFFE, U+EFFFF, U+FFFFE, U+FFFFF, U+10FFFE, or U+10FFFF
      */
     private static function renderAttribute(string $name, string $encodedValue = '', string $quote = '"'): string
     {
+        if ($name === '') {
+            throw new InvalidArgumentException('Attribute name is empty.');
+        }
+
+        if (
```
Comment (samdark) [inline src/Html.php:1981]: That might affect performance, but overall is a good thing.

**Review by samdark — APPROVED** [review]

**Pull request merged**